File: src/dashboard/ui/js/results/formatting.js
/**
 * Formatting utilities for evaluation results.
 *
 * This module provides functions used to format result values
 * and safely insert dynamic data into generated HTML.
 */

/**
 * Escapes HTML-sensitive characters.
 *
 * This prevents values received from the backend from being
 * interpreted as HTML markup when inserted into the page.
 *
 * @param {*} value
 * @returns {string}
 */
export function escapeHtml(value) {
    return String(value)
        .replace(/&/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;")
        .replace(/"/g, "&quot;")
        .replace(/'/g, "&#39;");
}

/**
 * Formats a cryptographic mode for display.
 *
 * @param {string} mode
 * @returns {string}
 */
export function formatMode(mode) {
    if (mode === "post_quantum") {
        return "Post-Quantum";
    }

    if (mode === "classical") {
        return "Classical";
    }

    return escapeHtml(mode);
}

/**
 * Formats an evaluation operation for display.
 *
 * @param {string} operation
 * @returns {string}
 */
export function formatOperation(operation) {
    if (!operation) {
        return "-";
    }

    const formatted =
        operation.charAt(0).toUpperCase()
        + operation.slice(1);

    return escapeHtml(formatted);
}

/**
 * Formats a date value using the English locale.
 *
 * @param {string} value
 * @returns {string}
 */
export function formatDate(value) {
    return new Date(value).toLocaleString(
        "en-GB"
    );
}

/**
 * Converts a byte value into a human-readable representation.
 *
 * @param {number} bytes
 * @returns {string}
 */
export function formatBytes(bytes) {
    if (bytes < 1024) {
        return `${Math.round(bytes)} B`;
    }

    if (bytes < 1024 ** 2) {
        return `${(
            bytes / 1024
        ).toFixed(2)} KB`;
    }

    if (bytes < 1024 ** 3) {
        return `${(
            bytes / 1024 ** 2
        ).toFixed(2)} MB`;
    }

    return `${(
        bytes / 1024 ** 3
    ).toFixed(2)} GB`;
}

/**
 * Formats a duration in milliseconds.
 *
 * @param {number} milliseconds
 * @returns {string}
 */
export function formatDuration(milliseconds) {
    if (milliseconds < 1000) {
        return `${milliseconds.toFixed(2)} ms`;
    }

    return `${(
        milliseconds / 1000
    ).toFixed(2)} s`;
}

File: src/dashboard/services/results_report/formatting.rs
//! Formatting utilities for evaluation reports.
//!
//! This module contains helper functions used to format values
//! and safely insert evaluation data into the generated HTML.

/// Adds a single derived metric to the HTML report.
///
/// If the metric is not available, a dash is displayed instead.
pub fn add_metric_cell(html: &mut String, title: &str, value: Option<f64>) {
    let value_text = match value {
        Some(value) => {
            format!("{value:.2} %")
        }
        None => "-".to_string(),
    };

    html.push_str(&format!(
        r#"<td class="metric-td">
<div class="metric-title">{}</div>
<div class="metric-value">{}</div>
</td>"#,
        escape_html(title),
        value_text,
    ));
}

/// Converts a byte value into a human-readable representation.
///
/// Values are displayed using bytes, kilobytes, megabytes
/// or gigabytes depending on their magnitude.
pub fn format_bytes(bytes: f64) -> String {
    if bytes < 1024.0 {
        return format!("{bytes:.0} B");
    }

    if bytes < 1024.0 * 1024.0 {
        return format!("{:.2} KB", bytes / 1024.0);
    }

    if bytes < 1024.0 * 1024.0 * 1024.0 {
        return format!("{:.2} MB", bytes / (1024.0 * 1024.0));
    }

    format!("{:.2} GB", bytes / (1024.0 * 1024.0 * 1024.0))
}

/// Escapes HTML-sensitive characters.
///
/// This prevents values originating from evaluation data
/// from being interpreted as HTML markup when inserted
/// into the generated report.
pub fn escape_html(value: &str) -> String {
    value
        .replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&#39;")
}
